### ISSUE TITLE
'optional' Array with nested 'requires' params 

### ISSUE THREAD

**Issue by ipkes** [ruby-grape/grape#723]
Grape v 0.8.0 <br/>
<b>Headers:</b>
Content-Type: application/json 
Accept: \*/\*
Accept-Encoding: gzip,deflate,sdch
Accept-Language: en-US,en;q=0.8,ru;q=0.6 <br/>
<b>Validation</b>:
<pre>
params do
  requires :docs do
    requires :type, type: String
    optional :files do 
      requires :href, type: String
    end
  end
end
</pre>
<b>Package json(with xml same problem)</b>:
<pre>
{
  "docs": [
       {"type": "aaa"}, 
       {"type": "bbb", "files": [{"href": "some_href"}] }
  ]
}
</pre>
<b>Result</b>:
<pre>
&lt;error&gt;
  &lt;message&gt;docs[files][href] is missing&lt;/message&gt;
&lt;/error&gt;
</pre>
This
<pre>
{
  "docs": [
      {"type": "aaa"}
  ]
}
</pre>
and this:
<pre>
{
  "docs": [
     {"type": "bbb", "files": [{"href": "som_href"}] }
  ]
}
</pre>
Works fine!
Help plz! What I'm doing wrong? Sorry if subject duplicate.

**Comment by justfalter** [comment]
I think this needs to be re-opened. Optional Array and Hash groups do not validate properly when the required parameter is missing. See pull request #883 for specs.

**Comment by dblock** [comment]
I think you're right @justfalter, reopening.

**Comment by NeMO84** [comment]
Any update on this?

**Comment by dblock** [comment]
Nobody has submitted a fix, please contribute.

**Comment by pocman** [comment]
Hi,

I have a similar issue : 
```
requires :route, type: Array, allow_blank: false do
    optional :calendar, type: Array, allow_blank: false do
        requires :starts_at, type: Time, allow_blank: false
        requires :ends_at, type: Time, allow_blank: false
    end
end
```
`"route": [{},{}]` is valid.
`"route":[{},{"calendar": [{"starts_at": "2016-05-02T09:00:00+00:00","ends_at":"2016-05-02T12:00:00+00:00"}]}]` returns 
```
{
  "error": "route[0][calendar][0][starts_at] is missing, route[0][calendar][0][ends_at] is missing"
}
```

**Comment by ipkes** [comment]
Hi @pocman! Can you confirm that if change type of `starts_at` and `ends_at` from `Time` to something else, `String` for example, error will disappear? I think this is not the same issue.

**Comment by ipkes** [comment]
@dbclock for this issue I think we have conflict demands.
[For example](https://github.com/ruby-grape/grape/blob/master/spec/grape/validations_spec.rb#L641)
<pre>
      params do
        group :children, type: Array do
          requires :name
          group :parents, type: Array do
            requires :name
          end
        end
      end
      it 'errors when a parameter is not present' do
        put_with_json '/within_array', children: [
          { name: 'Jim', parents: [{}] },
          { name: 'Job', parents: [{ name: 'Joy' }] }
        ]
        expect(last_response.status).to eq(400)
        expect(last_response.body).to eq('children[0][parents][0][name] is missing')
      end
</pre>

I'm not sure this is right. And I'm not sure should be there a difference in behavior between group and optional Arrays

**Comment by pocman** [comment]
@ipkes Yes, I changed type of starts_at and ends_at from Time to String and the error will disappeared !

**Comment by dblock** [comment]
So looks like the `Time` problem was resolved by #1380, but the original issue was different, @justfalter re-raised https://github.com/ruby-grape/grape/pull/883 and that was fixed. The only thing left is https://github.com/ruby-grape/grape/pull/911 by @al2o3cr. If someone has time, I'd appreciate a rebased run of #911 to see what's failing and what's not.

**Comment by dblock** [comment]
I believe https://github.com/ruby-grape/grape/commit/9a083581204a492ac7153cbd4d05e385f4a94793 closes this fully.
